middleman_protocol/middleman_protocol/helpers.py

```diff
@@ -0,0 +1,81 @@
+import socket
+
+from mypy.types import Iterable
+
+from middleman_protocol.constants import ESCAPE_CHARACTER
+from middleman_protocol.constants import ESCAPE_SEQUENCES
+from middleman_protocol.constants import FRAME_PAYLOAD_LENGTH_STARTING_BYTE
+from middleman_protocol.constants import FRAME_PAYLOAD_STARTING_BYTE
+from middleman_protocol.constants import FRAME_SEPARATOR
+from middleman_protocol.constants import FRAME_SEPARATOR_LENGTH
+from middleman_protocol.constants import RECEIVE_BYTES_PER_LOOP
+from middleman_protocol.message import AbstractMiddlemanMessage
+
+
+def handle_frame_receive(connection: socket.socket, separator_escaping_expected: bool=False) -> Iterable[bytes]:
+    """
+    Helper generator for receiving serialized Middleman protocol messages.
+
+    Yields received bytes.
+    """
+    assert isinstance(connection, socket.socket)
+    assert isinstance(separator_escaping_expected, bool)
+
+    received_data = b''
+    header_plus_payload_length = None
+    # Used to mark how many bytes was received since last decoding, to avoid decoding twice same part of data.
+    decoding_buffer_length = 0
+
+    while True:
+        received_data += connection.recv(RECEIVE_BYTES_PER_LOOP)
+
+        # If separator is met, reset received data and restart.
+        if (
+            len(received_data) >= FRAME_SEPARATOR_LENGTH and
+            received_data[-FRAME_SEPARATOR_LENGTH:] == FRAME_SEPARATOR
+        ):
+            received_data = b''
+            header_plus_payload_length = None
+            decoding_buffer_length = 0
+            continue
+
+        # If encoding is expected to be used by data sender,
+        # and there is encoded sequence in last received data,
+        # decode it and set encoding buffer.
+        if (
+            separator_escaping_expected and (
+                ESCAPE_SEQUENCES[FRAME_SEPARATOR] in received_data[decoding_buffer_length:] or
+                ESCAPE_SEQUENCES[ESCAPE_CHARACTER] in received_data[decoding_buffer_length:]
+            )
+        ):
+            received_data = (
+                received_data[:decoding_buffer_length] +
+                AbstractMiddlemanMessage.escape_decode_raw_message(received_data[decoding_buffer_length:])
+            )
```

Comment:
What if it's not a valid escape sequence? I.e. it starts with the escape character but cannot be converted to anything?

We could just leave it as is but I think it's better to treat it as an error because then later adding new sequences would change meaning of some existing ones.

Reply:
What do you mean by `it's not a valid escape sequence` ? There is an `if` above which checks that.

Reply:
> What do you mean by `it's not a valid escape sequence` ? There is an `if` above which checks that.

I'm not sure if this `if` is still there but I meant a sequence that cannot be converted to anything. E.g. when you use `FF` as the escape character and  define `FF 00` and `FF 01` as escape sequences then for example `FF 02`, `FF FF` or `FF` (as the last character in the stream, not followed by anything) are invalid sequences because we can't translate them.

Reply:
Right, it was not covered then but is with new implementation. :+1: done